src/cpp/ext/gsm/metadata_exchange.cc

```diff
@@ -120,6 +122,116 @@ absl::string_view GetStringValueFromUpbStruct(google_protobuf_Struct* struct_pb,
+  size_t pos_ = 0;
+  const std::vector<std::pair<absl::string_view, std::string>>& labels_;
+};
+
+class PeerLabelsIterable : public LabelsIterable {
+ public:
+  explicit PeerLabelsIterable(grpc_core::Slice remote_metadata) {
+    std::string decoded_metadata;
+    bool metadata_decoded = absl::Base64Unescape(
+        remote_metadata.as_string_view(), &decoded_metadata);
+    if (metadata_decoded) {
+      struct_pb_ = google_protobuf_Struct_parse(
+          decoded_metadata.c_str(), decoded_metadata.size(), arena_.ptr());
+      type_ = GetStringValueFromUpbStruct(struct_pb_, kMetadataExchangeTypeKey,
+                                          arena_.ptr());
+    }
+  }
+
+  absl::optional<std::pair<absl::string_view, absl::string_view>> Next()
+      override {
+    if (struct_pb_ == nullptr) {
+      return absl::nullopt;
+    }
+    if (pos_++ == 0) {
+      return std::make_pair(kPeerTypeAttribute, type_);
+    }
+    // Only handle GKE type for now.
+    if (type_ != kGkeType) {
+      return absl::nullopt;
+    }
+    switch (pos_ - 1) {
+      case 1:
+        return std::make_pair(
+            kPeerPodNameAttribute,
+            GetStringValueFromUpbStruct(struct_pb_, kMetadataExchangePodNameKey,
+                                        arena_.ptr()));
+      case 2:
+        return std::make_pair(
+            kPeerContainerNameAttribute,
+            GetStringValueFromUpbStruct(
+                struct_pb_, kMetadataExchangeContainerNameKey, arena_.ptr()));
+      case 3:
+        return std::make_pair(
+            kPeerNamespaceNameAttribute,
+            GetStringValueFromUpbStruct(
+                struct_pb_, kMetadataExchangeNamespaceNameKey, arena_.ptr()));
+      case 4:
+        return std::make_pair(
+            kPeerClusterNameAttribute,
+            GetStringValueFromUpbStruct(
+                struct_pb_, kMetadataExchangeClusterNameKey, arena_.ptr()));
+      case 5:
+        return std::make_pair(
+            kPeerLocationAttribute,
+            GetStringValueFromUpbStruct(
+                struct_pb_, kMetadataExchangeLocationKey, arena_.ptr()));
+      case 6:
+        return std::make_pair(
+            kPeerProjectIdAttribute,
+            GetStringValueFromUpbStruct(
+                struct_pb_, kMetadataExchangeProjectIdKey, arena_.ptr()));
+      case 7:
+        return std::make_pair(
+            kPeerCanonicalServiceAttribute,
+            GetStringValueFromUpbStruct(struct_pb_,
+                                        kMetadataExchangeCanonicalServiceKey,
+                                        arena_.ptr()));
+      default:
+        return absl::nullopt;
+    }
+  }
+
+  size_t Size() const override {
+    if (struct_pb_ == nullptr) {
+      return 0;
+    }
+    if (type_ != kGkeType) {
+      return 1;
+    }
+    return 8;
+  }
+
+  void ResetIteratorPosition() override { pos_ = 0; }
+
+ private:
+  upb::Arena arena_;
+  google_protobuf_Struct* struct_pb_ = nullptr;
+  absl::string_view type_;
```

Comment:
To save a bit of memory, this could be an enum.  That would also make the conditional more efficient in the future when we add support for GCE and other environments.

Reply:
done